crates/pathfinder/src/cairo/ext_py/sub_process.rs

```diff
@@ -0,0 +1,318 @@
+//! Launching and communication with the subprocess
+
+use super::{
+    de::{ChildResponse, RefinedChildResponse, Status},
+    ser::ChildCommand,
+    CallFailure, Command, SharedReceiver, SubProcessEvent, SubprocessError, SubprocessExitReason,
+};
+use anyhow::Context;
+use std::path::PathBuf;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{Child, ChildStdin, ChildStdout};
+use tokio::sync::{broadcast, mpsc};
+
+/// Launches a python subprocess, and executes calls on it until shutdown is initiated.
+///
+/// If the python process is killed, reaping it and restarting new one is handled by [`super::start`],
+/// similarly to spawning this as a task usually handled.
+///
+/// Launching happens in two stages, similar to the python process. Initially we only launch, then
+/// read `"ready\n"` from the subprocess and after that enter the loop where we contend for the
+/// commands.
+pub(super) async fn launch_python(
+    database_path: PathBuf,
+    commands: SharedReceiver<Command>,
+    status_updates: mpsc::Sender<SubProcessEvent>,
+    mut shutdown_rx: broadcast::Receiver<()>,
+) -> Option<(u32, Option<std::process::ExitStatus>, SubprocessExitReason)> {
+    let (mut child, pid, mut input, mut output, mut buffer) = match spawn(database_path).await {
+        Ok(tuple) => tuple,
+        Err(e) => {
+            let e = e.context("Failed to start python subprocess");
+            let _ = status_updates.send(SubProcessEvent::Failure(None, e)).await;
+            return None;
+        }
+    };
+
+    if status_updates
+        .send(SubProcessEvent::ProcessLaunched(pid))
+        .await
+        .is_err()
+    {
+        drop(input);
+        return None;
+    }
+
+    let mut command_buffer = Vec::new();
+
+    let exit_reason = loop {
```

Comment:
Was thinking of documenting why does this exit instead of having two loops. I think going through the `start` is needed either way for panics but ... otherwise, not so sure. It would at least ever increase the cognitive burden here at the benefit of reusing some of the buffers.

Reply:
I'll leave this to later thinking.

Reply:
Noting a comment here.